File: 2024 April/Delete Middle of Linked List/main.py
# Link: https://www.geeksforgeeks.org/problems/delete-middle-of-linked-list/1


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Llist:
    def __init__(self):
        self.head = None

    def insert(self, data, tail):
        node = Node(data)

        if not self.head:
            self.head = node
            return node

        tail.next = node
        return node


def printList(head):
    while head:
        print(head.data, end=" ")
        head = head.next
    print()


class Solution:
    def deleteMid(self, head):
        if not head.next:
            return
        s, f = head, head.next
        while f != None and f.next and f.next.next:
            f = f.next.next
            s = s.next
        s.next = s.next.next
        return head


if __name__ == "__main__":
    obj = Solution()
    values = [1, 2, 3, 4, 5]
    llist = Llist()
    tail = None
    for value in values:
        tail = llist.insert(value, tail)
    printList(obj.deleteMid(llist.head))

    values = [2, 4, 6, 7, 5, 1]
    llist = Llist()
    tail = None
    for value in values:
        tail = llist.insert(value, tail)
    printList(obj.deleteMid(llist.head))

File: 2024 April/Delete Middle of Linked List/main.cpp
// Link: https://www.geeksforgeeks.org/problems/delete-middle-of-linked-list/1

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

class Solution
{
public:
    Node *deleteMid(Node *head)
    {
        if (head == NULL or head->next == NULL)
            return NULL;
        int n = 0;
        Node *temp = head;
        while (temp)
        {
            temp = temp->next;
            n++;
        }
        n /= 2;
        Node *prev = head, *curr = head;
        while (n--)
        {
            prev = curr;
            curr = curr->next;
        }
        prev->next = curr->next;
        return head;
    }
};

int main()
{
    Solution obj;
    vector<int> lst = {1, 2, 3, 4, 5};
    Node *head = new Node(lst[0]);
    Node *temp = head;
    for (int i = 1; i < lst.size(); i++)
    {
        temp->next = new Node(lst[i]);
        temp = temp->next;
    }
    head = obj.deleteMid(head);
    while (head)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;

    lst = {2, 4, 6, 7, 5, 1};
    head = new Node(lst[0]);
    temp = head;
    for (int i = 1; i < lst.size(); i++)
    {
        temp->next = new Node(lst[i]);
        temp = temp->next;
    }
    head = obj.deleteMid(head);
    while (head)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
    return 0;
}